File: src/UE4_Assets/FByteBulkData.java
/**
 * 
 */
package UE4_Assets;

import UE4.FArchive;

/**
 * @author FunGames
 *
 */
public class FByteBulkData {
	private FByteBulkDataHeader header;
	private byte[] data;

	public FByteBulkData(FArchive Ar) throws ReadException {
		header = new FByteBulkDataHeader(Ar);
		int bulkDataFlags = header.getBulkDataFlags();
		data = new byte[header.getElementCount()];
		if(EBulkData.check(EBulkData.BULKDATA_Unused, bulkDataFlags)) {
			System.out.println("WARNING: Bulk with no data");
		} 
		//UE4
		else if(EBulkData.check((EBulkData.BULKDATA_OptionalPayload | EBulkData.BULKDATA_PayloadInSeperateFile), bulkDataFlags)) {
			//System.out.println(String.format("bulk data in %s file (flags=0x%X, pos=%d, size=%d)", EBulkData.check(EBulkData.BULKDATA_OptionalPayload, bulkDataFlags) ? ".uptnl" : ".ubulk", bulkDataFlags, header.getOffsetInFile() + Ar.getUbulkOffset(), header.getSizeOnDisk()));
			
			//Ubulk
			if(EBulkData.check(EBulkData.BULKDATA_PayloadInSeperateFile, bulkDataFlags)) {
				FArchive Ar_ubulk = Ar.getPayload("UBULK");
				if(Ar_ubulk != null) {
					int offset = (int) (header.getOffsetInFile() + Ar.getUbulkOffset());
					Ar_ubulk.Seek(offset);
					for(int i=0;i<header.getElementCount();i++) {
						data[i] = Ar_ubulk.readUInt8();
					}
				} else {
					throw new ReadException("No ubulk specified for this package but needed for loading", Ar.Tell());
				}
			}
			//TODO UPTNL
		}
		else if(EBulkData.check(EBulkData.BULKDATA_PayloadAtEndOfFile, bulkDataFlags)) {
			// stored in the same file, but at different position
			// save archive position
			int savePos = Ar.Tell();
			int saveStopper = Ar.GetStopper();
			// seek to data block and read data
			if(header.getOffsetInFile() - Ar.getUexpOffset() + header.getElementCount() <= Ar.data.length) {
				Ar.SetStopper((int) (header.getOffsetInFile() + header.getElementCount()));
				Ar.Seek((int) header.getOffsetInFile());
				//System.out.println(String.format("bulk data in %s file (Payload at end of file) (flags=0x%X, pos=%d, size=%d)", ".uexp", bulkDataFlags, header.getOffsetInFile() - Ar.getUexpOffset(), header.getSizeOnDisk()));
				for(int i=0;i<header.getElementCount(); i++) {
					data[i] = Ar.readUInt8();
				}
			} else {
				throw new ReadException(String.format("Can't read %d bytes at offset %d, max index: %d", header.getElementCount(), header.getOffsetInFile() - Ar.getUexpOffset(), Ar.data.length), Ar.Tell());
			}
			// restore archive position
			Ar.Seek(savePos);
			Ar.SetStopper(saveStopper);	
		}
		else if(EBulkData.check(EBulkData.BULKDATA_ForceInlinePayload, bulkDataFlags)) {
			//System.out.println(String.format("bulk data in %s file (Force payload inline) (flags=0x%X, pos=%d, size=%d)", ".uexp", bulkDataFlags, Ar.Tell(), header.getSizeOnDisk()));
			for(int i=0; i<header.getElementCount(); i++) {
				data[i] = Ar.readUInt8();
			}
		}
	}

	public FByteBulkDataHeader getHeader() {
		return header;
	}

	public byte[] getData() {
		return data;
	}
	
	
}

File: src/main/kotlin/me/fungames/jfortniteparse/ue4/assets/objects/FByteBulkData.kt
package me.fungames.jfortniteparse.ue4.assets.objects

import me.fungames.jfortniteparse.LOG_JFP
import me.fungames.jfortniteparse.exceptions.ParserException
import me.fungames.jfortniteparse.ue4.assets.enums.EBulkDataFlags.*
import me.fungames.jfortniteparse.ue4.assets.reader.FAssetArchive
import me.fungames.jfortniteparse.ue4.assets.util.PayloadType
import me.fungames.jfortniteparse.ue4.assets.writer.FAssetArchiveWriter

class FByteBulkData {
    var header: FByteBulkDataHeader
    var data: ByteArray
    val isBulkDataLoaded get() = data.isNotEmpty()

    constructor(Ar: FAssetArchive) {
        header = FByteBulkDataHeader(Ar)
        val bulkDataFlags = header.bulkDataFlags
        data = ByteArray(header.elementCount.toInt())
        when {
            header.elementCount == 0L -> {
                // Nothing to do here
            }
            BULKDATA_Unused.check(bulkDataFlags) -> {
                LOG_JFP.warn("Bulk with no data")
            }
            BULKDATA_ForceInlinePayload.check(bulkDataFlags) -> {
                LOG_JFP.debug("bulk data in .uexp file (Force Inline Payload) (flags=$bulkDataFlags, pos=${header.offsetInFile}, size=${header.sizeOnDisk})")
                Ar.read(data)
            }
            BULKDATA_PayloadInSeperateFile.check(bulkDataFlags) -> {
                LOG_JFP.debug("bulk data in .ubulk file (Payload In Seperate File) (flags=$bulkDataFlags, pos=${header.offsetInFile}, size=${header.sizeOnDisk})")
                val ubulkAr = Ar.getPayload(when {
                    BULKDATA_OptionalPayload.check(bulkDataFlags) -> PayloadType.UPTNL
                    BULKDATA_MemoryMappedPayload.check(bulkDataFlags) -> PayloadType.M_UBULK
                    else -> PayloadType.UBULK
                })
                if (ubulkAr.size() > 0) {
                    ubulkAr.seek(header.offsetInFile.toInt())
                    ubulkAr.read(data)
                } else { // bulk data file not found
                    data = ByteArray(0)
                }
            }
            BULKDATA_PayloadAtEndOfFile.check(bulkDataFlags) -> {
                //stored in same file, but at different position
                //save archive position
                val savePos = Ar.pos()
                if (header.offsetInFile.toInt() + header.elementCount <= Ar.size()) {
                    Ar.seek(header.offsetInFile.toInt())
                    Ar.read(data)
                } else {
                    throw ParserException("Failed to read PayloadAtEndOfFile, ${header.offsetInFile} is out of range", Ar)
                }
                Ar.seek(savePos)
            }
        }
    }

    fun serialize(Ar: FAssetArchiveWriter) {
        val bulkDataFlags = header.bulkDataFlags
        when {
            BULKDATA_Unused.check(bulkDataFlags) -> {
                header.serialize(Ar)
            }
            BULKDATA_ForceInlinePayload.check(bulkDataFlags) -> {
                header.offsetInFile = (Ar.relativePos() + 28).toLong()
                header.elementCount = data.size.toLong()
                header.sizeOnDisk = data.size.toLong()
                header.serialize(Ar)
                Ar.write(data)
            }
            BULKDATA_PayloadInSeperateFile.check(bulkDataFlags) -> {
                val ubulkAr = Ar.getPayload(PayloadType.UBULK)
                header.offsetInFile = ubulkAr.relativePos().toLong()
                header.elementCount = data.size.toLong()
                header.sizeOnDisk = data.size.toLong()
                header.serialize(Ar)
                ubulkAr.write(data)
            }
            BULKDATA_OptionalPayload.check(bulkDataFlags) -> {
                val ubulkAr = Ar.getPayload(PayloadType.UPTNL)
                header.offsetInFile = ubulkAr.relativePos().toLong()
                header.elementCount = data.size.toLong()
                header.sizeOnDisk = data.size.toLong()
                header.serialize(Ar)
                ubulkAr.write(data)
            }
            else -> throw ParserException("Unsupported BulkData type $bulkDataFlags")
        }
    }

    constructor(header: FByteBulkDataHeader, data: ByteArray) {
        this.header = header
        this.data = data
    }
}